napari/_vispy/vispy_base_layer.py

```diff
@@ -192,10 +192,106 @@ def _reset_base(self):
+
+        Parameters
+        ----------
+        rounding_factor : int, optional
+            Use a rounding factor to prevent repeated triggering of requesting
+            of new pyramid tiles for small camera movements. The rounding
+            factor is in pixel until of the currently viewed tile. Floor and
+            ceil rounding operations are used appropriately to ensure that the
+            rounded requested field always exceeds the needed field to fill the
+            canvas. There is a tradeoff between larger rounding factors causing
+            less frequent updates and tile fetches, but bigger tiles. A factor
+            of around 50 gives good results empically.
+
+        Returns
+        ----------
+        corner_pixels : array
+            Coordinates of top left and bottom right canvas pixel in the data.
+        requested_shape : array
+            Shape of requested tile in data coordinates.
+        """
+        nd = self.layer.dims.ndisplay
+        # Find image coordinate of top left canvas pixel
+        if self.node.canvas is not None:
+            offset = self.translate[:nd] / self.scale[:nd]
+            tl_raw = np.floor(self._transform_position([0, 0]) + offset[::-1])
+            br_raw = np.ceil(
+                self._transform_position(self.node.canvas.size) + offset[::-1]
+            )
+        else:
+            tl_raw = [0] * nd
+            br_raw = [1] * nd
+
+        if rounding_factor is not None:
+            # Perform rounding to prevent repeated triggering of requesting
+            # of new pyramid tiles for small camera movements
+            tl_raw = rounding_factor * np.floor(
+                np.divide(tl_raw, rounding_factor)
+            )
+            br_raw = rounding_factor * np.ceil(
+                np.divide(br_raw, rounding_factor)
+            )
+
+        top_left = np.zeros(self.layer.ndim)
+        bottom_right = np.zeros(self.layer.ndim)
+        for d, tl, br in zip(self.layer.dims.displayed, tl_raw, br_raw):
+            top_left[d] = tl
+            bottom_right[d] = br
+
+        corner_pixels = np.array([top_left, bottom_right]).astype(int)
+
+        # Clip according to the max data of the level shape
+        corner_pixels = np.clip(
+            corner_pixels,
+            0,
+            np.subtract(self.layer.level_shapes[self.layer.data_level], 1),
+        )
+
+        # Scale to full resolution of the data
+        requested_shape = (
+            corner_pixels[1] - corner_pixels[0]
+        ) * self.layer.downsample_factors[self.layer.data_level]
+
+        return corner_pixels, requested_shape
+
     def on_draw(self, event):
         """Called whenever the canvas is drawn.
+
+        This is triggered from vispy whenever new data is sent to the canvas or
+        the camera is moved and is connected in the `QtViewer`.
         """
         self.layer.scale_factor = self.scale_factor
+        if self.layer.is_pyramid:
```

Comment:
put corner_pixels on base layer, and move this out of `is_pyramid` for @AhmetCanSolak

Reply:
so a potential problem now is that @AhmetCanSolak probably doesn't want the values to be clipped and rounded, but the pyramid code does. To solve this I will move the rounding, clipping code elsewhere - will clean some things up